keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,416 @@
+- [X] (R) Production readiness review completed
+- [X] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Many workloads benefit from knowing the underlying topology that a Pod is scheduled in.
+However, today there is no mechanism in Kubernetes to pass down node topology information to Pods.
+The KEP outlines how node topology labels will made available to Pods via the downward API.
+
+Exposing node topology labels through the downward API provides a native method
+from which to attain node label values in pods without the need of a sidecar
+to extract information.
+
+## Motivation
+
+We’d like to change the runtime behavior of containers based on node labels.
+In our case, we’re using a CNI with DaemonSets to perform network setup, and
+would like to configure the network differently based on the presence of a node
+label.
+
+A number of other use cases exist for providing node labels to pods. One
+example is utilizing topology data from cloud providers, which are automatically
+assigned to nodes, and then used to tailor workloads to minimize cross-zone data
+transfers and reduce costs. Having an easy way for pods to access these node
+topology labels would provide users a straightforward, maintainable way to
+optimize their workloads given topology constraints.
+
+While "topology" is usually associated with the physical layout of a cluster,
+it can also be used to describe other types of information about the cluster.
+This KEP proposes to allow the expansion of the concept of topology to include
+user-defined aspects about their cluster nodes, and in turn provide a way for
+pods to receive this information.
+
+Workarounds today typically involve using an initContainer to query the
+Kubernetes API and then pass data via shared volume to other containers within
+the same pod. By comparison, this proposal would reduce the number of service
+accounts and API server clients. Another workaround is to use webhooks to inject
+labels into pods, but this relies on advance knowledge of where the pod is going
+to be scheduled and requires the webhook to be running and available at the time
+of pod creation. This proposal would provide an easier way to access node labels
+from pods, and would be more efficient than the current workarounds.
+
+### Goals
+
+* Values from labels `topology.k8s.io/*` and `kubernetes.io/hostname` are made
```

Comment:
Is the goal inclusive of storing those labels in the APIserver, such that they are selectable like normal labels?  Or is the goal to ONLY publish via downward API and specifally NOT be selectable?  Or "don't care".  I think we should choose one explicitly.

Reply:
This was copy over from the old PR, but I propose we don't select against all labels and codify the topologies in downward API, something like this:
```
        - name: ZONE_ENV_VAR
          valueFrom:
            topologyRef:
              path: topology.zone
        - name: REGION_ENV_VAR
          valueFrom:
            topologyRef:
              path: topology.region
```

What do you think?